scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -28,16 +39,28 @@ sealed abstract class BeamOp[+A] {
 
   def flatMap[B](f: A => TraversableOnce[B])(implicit kryoCoder: KryoCoder): BeamOp[B] =
     parDo(FlatMapFn(f))
+}
 
-  def parDo[C >: A, B](f: DoFn[C, B])(implicit kryoCoder: KryoCoder): BeamOp[B] = {
-    val pTransform = new PTransform[PCollection[C], PCollection[B]]() {
-      override def expand(input: PCollection[C]): PCollection[B] = input.apply(ParDo.of(f))
+object BeamOp extends Serializable {
+  def planMapGroup[K, V, U](
+    pcoll: PCollection[KV[K, java.lang.Iterable[V]]],
+    reduceFn: (K, Iterator[V]) => Iterator[U])(implicit ordK: Ordering[K], kryoCoder: KryoCoder): PCollection[KV[K, java.lang.Iterable[U]]] = {
+    reduceFn match {
```

Comment:
by the way, my suggestion is that maybe Beam can do better optimization if we follow this pattern, but that could be incorrect... so it could be always doing what we are doing in `notComposedOrSum` is the optimal thing most of the time.

For instance by using `Combine.groupedValues` we are missing the chance to use `sa.sg.sumOption` which can be a huge win for some cases (since it can use mutability inside that method to speed up aggregating large groups)...

I don't know very much about Beam so I don't know the answer... this should probably be something that benchmarks answer the question of (presumably with cases where there are at least some examples where there are MANY values for each key).

Reply:
We plan to do some performance testing eventually. For instance trying to use `Combine.perKey` inside sumByLocalKeys. This adds a shuffle step for sure so in theory it shouldn't perform better. I am not sure if Dataflow runner can anyhow optimize the next shuffle.

Interface of CombineFn is similar to that of an accumulator so I would guess that `Combine.groupedValues` is doing some map side aggregation but we would need to verify it.